File: crates/dw-cli/src/guide.rs
use dw_ui::TerminalTheme;

pub(crate) fn print_guide(version: &str) {
    print_styled_lines(&render_guide(version, &TerminalTheme::stdout_auto()));
}

fn print_styled(line: &str) {
    println!("{}", TerminalTheme::stdout_auto().style_line(line, false));
}

fn print_styled_lines(lines: &[String]) {
    for line in lines {
        print_styled(line);
    }
}

fn render_guide(version: &str, theme: &TerminalTheme) -> Vec<String> {
    vec![
        theme.command(&format!("Dev Workflow {version}")),
        "Guide de démarrage pas à pas".into(),
        String::new(),
        "1. Vérifier l'installation".into(),
        format!("   {}", theme.command("dw version")),
        format!("   {}", theme.command("dw doctor")),
        "   Corriger les prérequis signalés avant de créer des workspaces.".into(),
        String::new(),
        "2. Initialiser le root DevWorkflow".into(),
        format!("   {}", theme.command("dw init")),
        format!("   {}", theme.command("dw config show")),
        "   Le root contient config, schemas, cache, projets, workspaces et contextes agents.".into(),
        "   Pour choisir un chemin explicite:".into(),
        format!("   {}", theme.command("dw init --root ~/dev/dw")),
        String::new(),
        "3. Brancher Azure DevOps".into(),
        format!("   {}", theme.command("dw auth login")),
        format!("   {}", theme.command("dw auth status")),
        format!("   {}", theme.command("dw ado assigned")),
        "   Sans --project, dw propose les projets configurés quand le terminal est interactif.".into(),
        String::new(),
        "4. Créer un workspace de travail".into(),
        format!("   {}", theme.command("dw task start <work-item-id>")),
        "   Sans --execute, dw affiche le plan: branche, repositories, worktrees, handoffs.".into(),
        format!("   {}", theme.command("dw task start <work-item-id> --execute")),
        format!("   {}", theme.command("dw task open --continue")),
        "   L'agent configuré s'ouvre dans le workspace avec le contexte DevWorkflow.".into(),
        String::new(),
        "5. Boucle quotidienne".into(),
        format!("   {}", theme.command("dw task status")),
        format!("   {}", theme.command("dw task list")),
        format!("   {}", theme.command("dw task current")),
        format!("   {}", theme.command("dw task preflight --continue")),
        format!("   {}", theme.command("dw task sync --continue")),
        "   Utiliser preflight avant l'implémentation et sync pour rafraîchir task.json depuis ADO.".into(),
        String::new(),
        "6. Gérer le contenu du workspace".into(),
        format!("   {}", theme.command("dw task add-work-item --continue")),
        format!("   {}", theme.command("dw task remove-work-item --continue")),
        format!("   {}", theme.command("dw task add-repo --continue")),
        format!("   {}", theme.command("dw task repo-latest --continue")),
        "   Les commandes interactives proposent les valeurs locales quand elles sont disponibles.".into(),
        String::new(),
        "7. Préparer la fin de tâche".into(),
        format!("   {}", theme.command("dw task handoff-validate --continue")),
        format!("   {}", theme.command("dw task commit --continue")),
        format!("   {}", theme.command("dw task finish --continue")),
        "   Ces commandes sont en preview par défaut. Ajouter --execute seulement après lecture du plan.".into(),
        format!("   {}", theme.command("dw task finish --continue --execute")),
        String::new(),
        "8. Nettoyer".into(),
        format!("   {}", theme.command("dw task teardown --continue")),
        format!("   {}", theme.command("dw task prune")),
        "   teardown et prune suppriment seulement avec --execute, et demandent confirmation en interactif.".into(),
        String::new(),
        "9. ADO, DB et contexte IA".into(),
        format!("   {}", theme.command("dw ado work-item <id>")),
        format!("   {}", theme.command("dw ado context <id>")),
        format!("   {}", theme.command("dw ado changelog <ids>")),
        format!("   {}", theme.command("dw db schema")),
        format!("   {}", theme.command("dw db describe <table>")),
        format!("   {}", theme.command("dw db query --sql \"select top 20 * from ...\"")),
        format!("   {}", theme.command("dw agent context")),
        "   Les accès DB sont protégés par la garde read-only.".into(),
        String::new(),
        "10. Productivité shell".into(),
        format!("   {}", theme.command("dw completion show")),
        format!("   {}", theme.command("dw completion install fish")),
        format!("   {}", theme.command("dw completion install powershell")),
        "   Les complétions proposent options, projets, repositories, workspaces, databases et descriptions.".into(),
        String::new(),
        "Diagnostic rapide".into(),
        format!("   {}", theme.command("dw doctor --fix")),
        format!("   {}", theme.command("dw config doctor")),
        format!("   {}", theme.command("dw refresh")),
        "   refresh régénère schemas et contextes agents sans écraser les fichiers utilisateur.".into(),
    ]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn guide_renders_version_and_next_steps() {
        let lines = render_guide("2026.07.02.3+54011f0", &TerminalTheme::plain());

        assert_eq!(lines[0], "Dev Workflow 2026.07.02.3+54011f0");
        assert!(lines.contains(&"Guide de démarrage pas à pas".into()));
        assert!(lines.iter().any(|line| line.contains("dw init")));
        assert!(lines.iter().any(|line| line.contains("dw doctor")));
        assert!(
            lines
                .iter()
                .any(|line| line.contains("dw task start <work-item-id>"))
        );
        assert!(
            lines
                .iter()
                .any(|line| line.contains("dw task finish --continue"))
        );
        assert!(lines.iter().any(|line| line.contains("dw db query")));
        assert!(lines.iter().any(|line| line.contains("dw ado assigned")));
        assert!(lines.iter().any(|line| line.contains("dw completion show")));
    }
}

File: internal/console/guide.go
package console

import (
	"strings"

	"github.com/sachahjkl/dw/internal/l10n"
)

type GuideResult struct{ Version string }

type GuideStep struct {
	Title    MessageID
	Commands []string
	Detail   MessageID
}

var guideSteps = []GuideStep{
	{"guide.step.installation", []string{"dw version", "dw doctor"}, "guide.step.installation.detail"},
	{"guide.step.initialize", []string{"dw init", "dw config show", "dw init --root ~/dev/dw"}, "guide.step.initialize.detail"},
	{"guide.step.ado", []string{"dw provider auth login azure-devops", "dw provider auth status azure-devops", "dw work item list --project <project>"}, "guide.step.ado.detail"},
	{"guide.step.workspace", []string{"dw workspace start <work-item-id>", "dw workspace start <work-item-id> --execute", "dw workspace open --continue"}, "guide.step.workspace.detail"},
	{"guide.step.daily", []string{"dw workspace status", "dw workspace list", "dw workspace current", "dw workspace preflight --continue", "dw workspace sync --continue"}, "guide.step.daily.detail"},
	{"guide.step.contents", []string{"dw workspace item add --continue", "dw workspace item remove --continue", "dw workspace repo add <repo>", "dw workspace repo latest --continue"}, "guide.step.contents.detail"},
	{"guide.step.complete", []string{"dw workspace handoff validate --continue", "dw workspace commit --continue", "dw workspace finish --continue", "dw workspace finish --continue --execute"}, "guide.step.complete.detail"},
	{"guide.step.cleanup", []string{"dw workspace teardown --continue", "dw workspace prune"}, "guide.step.cleanup.detail"},
	{"guide.step.tools", []string{"dw work item show <id>", "dw work context show <id>", "dw work changelog <ids>", "dw data catalog", "dw data describe <resource>", "dw data query --query \"select top 20 * from ...\"", "dw agent context"}, "guide.step.tools.detail"},
	{"guide.step.completion", []string{"dw completion show", "dw completion install fish", "dw completion install powershell"}, "guide.step.completion.detail"},
}

func RenderVersion(version string, localizer Localizer, theme Theme) Output {
	localizer = WithConsoleMessages(localizer)
	return TextOutput(FormatHuman, theme.Title(localize(localizer, "build.product")+" "+version))
}

func RenderGuide(result GuideResult, localizer Localizer, theme Theme) string {
	localizer = WithConsoleMessages(localizer)
	blocks := []string{
		theme.Title(localize(localizer, "guide.title", l10n.A("version", result.Version))),
		localize(localizer, "guide.subtitle"),
	}
	for index, step := range guideSteps {
		lines := []string{theme.Label(localize(localizer, "guide.step.numbered", l10n.A("number", index+1), l10n.A("title", localize(localizer, step.Title))))}
		for _, command := range step.Commands {
			lines = append(lines, "  "+theme.Command(command))
		}
		lines = append(lines, "  "+localize(localizer, step.Detail))
		blocks = append(blocks, strings.Join(lines, "\n"))
	}
	blocks = append(blocks, theme.Panel(strings.Join([]string{
		theme.Label(localize(localizer, "guide.diagnostics")),
		"  " + theme.Command("dw doctor --fix"),
		"  " + theme.Command("dw config doctor"),
		"  " + theme.Command("dw refresh"),
		"  " + localize(localizer, "guide.diagnostics.detail"),
	}, "\n")))
	return strings.Join(blocks, "\n\n")
}
